src/crypto/slow-hash.c

```diff
@@ -59,35 +60,104 @@ extern int aesb_pseudo_round(const uint8_t *in, uint8_t *out, const uint8_t *exp
   } while(0)
 
 #define VARIANT1_2(p) \
-  do if (variant > 0) \
+  do if (variant == 1) \
   { \
     xor64(p, tweak1_2); \
   } while(0)
 
 #define VARIANT1_CHECK() \
   do if (length < 43) \
   { \
-    fprintf(stderr, "Cryptonight variants need at least 43 bytes of data"); \
+    fprintf(stderr, "Cryptonight variant 1 needs at least 43 bytes of data"); \
     _exit(1); \
   } while(0)
 
 #define NONCE_POINTER (((const uint8_t*)data)+35)
 
 #define VARIANT1_PORTABLE_INIT() \
   uint8_t tweak1_2[8]; \
-  do if (variant > 0) \
+  do if (variant == 1) \
   { \
     VARIANT1_CHECK(); \
     memcpy(&tweak1_2, &state.hs.b[192], sizeof(tweak1_2)); \
     xor64(tweak1_2, NONCE_POINTER); \
   } while(0)
 
 #define VARIANT1_INIT64() \
-  if (variant > 0) \
+  if (variant == 1) \
   { \
     VARIANT1_CHECK(); \
   } \
-  const uint64_t tweak1_2 = variant > 0 ? (state.hs.w[24] ^ (*((const uint64_t*)NONCE_POINTER))) : 0
+  const uint64_t tweak1_2 = (variant == 1) ? (state.hs.w[24] ^ (*((const uint64_t*)NONCE_POINTER))) : 0
+
+#define VARIANT2_INIT() \
+  uint32_t division_result[2]; \
+  uint32_t sqrt_result; \
+  do if (variant == 2) \
+  { \
+    *((uint64_t*)division_result) = 0; \
+    sqrt_result = 0; \
+  } while (0)
+
+#define VARIANT2_SHUFFLE_SSE2(base_ptr, offset) \
+  do if (variant == 2) \
+  { \
+    const __m128i chunk1 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x10))); \
+    const __m128i chunk2 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x20))); \
+    const __m128i chunk3 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x30))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x10)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk3, _MM_SHUFFLE(2, 0, 3, 1)), _MM_SHUFFLE(3, 1, 2, 0))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x20)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk1, _MM_SHUFFLE(0, 3, 1, 2)), _MM_SHUFFLE(0, 3, 1, 2))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x30)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk2, _MM_SHUFFLE(3, 1, 2, 0)), _MM_SHUFFLE(2, 0, 3, 1))); \
+  } while (0)
+
+#define VARIANT2_SHUFFLE_PORTABLE(base_ptr, offset) \
+  do if (variant == 2) \
+  { \
+    uint32_t* chunk1 = (uint32_t*)((base_ptr) + ((offset) ^ 0x10)); \
```

Comment:
This isn't portable. This is an aliasing violation that can result in misaligned access errors on some CPUs.

Reply:
base_ptr is actually a scratchpad pointer here. It's allocated by malloc/VirtualAlloc/mmap. It will be 8-byte aligned in the worst case. offset is always a multiple of 16, so misaligned access is not possible here.

Reply:
Yikes, its reliant on the code below. This is still an aliasing violation, but AFAIK will work on our targeted compilers.

Reply:
It's safe as long as compiled code passes all tests. This PR has test vectors to catch any bug.